cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu

```diff
@@ -0,0 +1,238 @@
+{
+  const int num_items = 1;
+  thrust::device_vector<int> in(num_items, 42);
+  thrust::device_vector<int> out_num_runs(1, 42);
+
+  // Note intentionally no discard_iterator as we want to ensure nothing is written to the output arrays
+  run_length_encode(in.begin(),
+                    static_cast<cuda::std::size_t*>(nullptr),
+                    static_cast<std::uint16_t*>(nullptr),
+                    out_num_runs.begin(),
+                    num_items);
+
+  REQUIRE(out_num_runs.front() == 0);
+}
+
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns can handle all unique", "[device][run_length_encode]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = 10;
+  thrust::device_vector<int> out_num_runs(1, -1);
+
+  run_length_encode(thrust::make_counting_iterator(type{}),
+                    static_cast<int*>(nullptr),
+                    static_cast<int*>(nullptr),
+                    out_num_runs.begin(),
+                    num_items);
+
+  REQUIRE(out_num_runs.front() == 0);
+}
+
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns can handle all equal", "[device][run_length_encode]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = 10;
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<int>  out_offsets(1, -1);
+  thrust::device_vector<int>  out_lengths(1, -1);
+  thrust::device_vector<int>  out_num_runs(1, -1);
+  c2h::gen(CUB_SEED(2), in);
+  thrust::fill(in.begin(), in.end(), in.front());
+
+  run_length_encode(in.begin(),
+                    out_offsets.begin(),
+                    out_lengths.begin(),
+                    out_num_runs.begin(),
+                    num_items);
+
+  REQUIRE(out_offsets.front()  == 0);
+  REQUIRE(out_lengths.front()  == num_items);
+  REQUIRE(out_num_runs.front() == 1);
+}
+
+template<class T, class Index>
+bool validate_results(const thrust::device_vector<T>&     in,
+                      const thrust::device_vector<Index>& out_offsets,
+                      const thrust::device_vector<Index>& out_lengths,
+                      const thrust::device_vector<Index>& out_num_runs,
+                      const int num_items)
+{
+  for(cuda::std::size_t run = 0; run < static_cast<cuda::std::size_t>(out_num_runs.front()); ++run) {
+    const cuda::std::size_t first_index = static_cast<cuda::std::size_t>(out_offsets[run]);
+    const cuda::std::size_t final_index = first_index + static_cast<cuda::std::size_t>(out_lengths[run]);
+
+    // Ensure we started a new run
+    if (first_index > 0) {
+      if (in[first_index] == in[first_index - 1]) {
+        return false;
+      }
+    }
+
+    // Ensure the run is valid
+    for (cuda::std::size_t running_index = first_index + 1; running_index < final_index; ++running_index) {
+      if (!(in[first_index] == in[running_index])) {
+        return false;
+      }
+    }
```

Comment:
important: I don't think it's our code style, please apply clang-format

Reply:
Are you talking about the !( == ) This is because the custom type only knows ==

Reply:
I'm talking about `if () {` instead of `if () \n {`

Reply:
![angry-cat](https://github.com/NVIDIA/cccl/assets/3017120/7af7c3c8-4131-4619-9c74-aa78718d712c)